pkg/kube/client/ssa/caching_unstructured_extractor.go

```diff
@@ -0,0 +1,237 @@
+// SPDX-FileCopyrightText: 2024 The Crossplane Authors <https://crossplane.io>
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package ssa
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/managedfields"
+	applymetav1 "k8s.io/client-go/applyconfigurations/meta/v1"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"k8s.io/kube-openapi/pkg/handler3"
+	"k8s.io/kube-openapi/pkg/schemamutation"
+	"k8s.io/kube-openapi/pkg/spec3"
+	"k8s.io/kube-openapi/pkg/validation/spec"
+)
+
+// cachingUnstructuredExtractor is a caching implementation of v1.UnstructuredExtractor
+// using OpenAPI V3 discovery information.
+// TODO(erhan): try to upstream this code in kubernetes
+type cachingUnstructuredExtractor struct {
+	// added as field to not break the interface for other funcs, instantiated at each reconcile
+	ctx   context.Context
+	cache *GVKParserCache
+	dc    discovery.DiscoveryInterface
+}
+
+// NewCachingUnstructuredExtractor returns a new cachingUnstructuredExtractor
+func NewCachingUnstructuredExtractor(ctx context.Context, dc discovery.DiscoveryInterface, cache *GVKParserCache) (applymetav1.UnstructuredExtractor, error) {
+	return &cachingUnstructuredExtractor{
+		dc:    dc,
+		cache: cache,
+		ctx:   ctx,
+	}, nil
+}
+
+// Extract extracts the applied configuration owned by fieldManager from an unstructured object.
+// Note that the apply configuration itself is also an unstructured object.
+func (e *cachingUnstructuredExtractor) Extract(object *unstructured.Unstructured, fieldManager string) (*unstructured.Unstructured, error) {
+	return e.extractUnstructured(object, fieldManager, "")
+}
+
+// ExtractStatus is the same as ExtractUnstructured except
+// that it extracts the status subresource applied configuration.
+// Experimental!
+func (e *cachingUnstructuredExtractor) ExtractStatus(object *unstructured.Unstructured, fieldManager string) (*unstructured.Unstructured, error) {
+	return e.extractUnstructured(object, fieldManager, "status")
+}
+
+func discoveryPaths(ctx context.Context, rc rest.Interface) (map[string]OpenAPIGroupVersion, error) {
+	data, err := rc.Get().
+		AbsPath("/openapi/v3").
+		Do(ctx).
+		Raw()
+
+	if err != nil {
+		return nil, err
+	}
+
+	discoMap := &handler3.OpenAPIV3Discovery{}
+	err = json.Unmarshal(data, discoMap)
+	if err != nil {
+		return nil, err
+	}
+
+	oapiPathsToGV := map[string]OpenAPIGroupVersion{}
+	for path, oapiGV := range discoMap.Paths {
+		parse, err := url.Parse(oapiGV.ServerRelativeURL)
+		if err != nil {
+			return nil, err
+		}
+		useClientPrefix := strings.HasPrefix(oapiGV.ServerRelativeURL, "/openapi/v3")
+		etag := parse.Query().Get("hash")
```

Comment:
is hash considered an interface in Kube ? Can it be renamed to etag tomorrow?

Reply:
not an interface in exact terms, but the upstream [implementation](https://github.com/kubernetes/kube-openapi/blame/9e1beecbcb384a484fb5cbd41024f316d51fdfac/pkg/handler3/handler.go#L128) and [comment](https://github.com/kubernetes/kube-openapi/blame/9e1beecbcb384a484fb5cbd41024f316d51fdfac/pkg/handler3/handler.go#L58) seems to imply that way. Anyways, this implementation in general is hopefully will be an interim solution.

Reply:
Upstream is built in a way that the client does not need to know the path format. It might be that we get broken here. Chance is not super high, but possible. Alternative: just store the full path with query params.

Reply:
agreed on the concern, but for the alternative, AFAIU if the server is changed in a way that it does not return any kind of hash-like query param, we risk unintentionally caching with a non-unique key, therefore operating on stale schema.

In the current state of the PR, we'll skip caching and we will only lose the caching ability, in case we cannot find `hash` in query params, but we continue operating with correct schema.

WDTY about an follow-up issue for an integration test implementation that checks whether discovery returns `hash`?

Reply:
Isn't that just a 5 liner? Just do it.